packages/platform-core/src/object_accumulator.ts

```diff
@@ -0,0 +1,49 @@
+import _ from 'lodash';
+
+/**
+ * This error is thrown when there's a collision in
+ */
+export class ObjectAccumulatorPropertyAlreadyExistsError extends Error {
+  /**
+   * Creates property already exists error.
+   */
+  constructor(
+    readonly key: string,
+    readonly existingValue: unknown,
+    readonly incomingValue: unknown
+  ) {
+    super(`Property ${key} already exists`);
+  }
+}
+
+/**
+ * A class that can accumulate (squash merge) objects into single instance.
+ */
+export class ObjectAccumulator<T> {
```

Comment:
This ended up here as generic component because the logic of accumulating client config partials at synthesis and config generation is pretty much the same. The different part is error message (achieved by translating the error thrown here).

Alternatives considered:
1. `class ClientConfigAccumulator` in `client-config` package (used in both client config and backend). I didn't like this idea because it would increase area of  `client-config` public api (which is intended to be customer facing).
2. Duplicate this logic between `backend` and `client-config`